# SV Analysis Scripts ReadMe
## Table of contents
- [Scripts for General Usage](#Scripts-for-General-Usage)
- [SV Caller Post Processing](#SV-Caller-Post-Processing)
- [Basic Analysis](#Basic-Analysis)

# Scripts for General Usage
## vcf_filter.py
**location:** `General_Usage/vcf_filter.py`
### Usage
Filtering the VCF file with certain criteria (see **Parameters** section) for Truvari and Hap-eval evaluation
### Inputs/Outputs
**Input:** VCF file

**Output:** 3 VCF files for INS/DEL/INS&DEL
### Parameters
```
--vcf,-v: Your input VCF file
--out_dir,-o_dir Directory where the outputs will be stored
--prefix,-p [Optional] Prefix of the output VCFs, if not given, will extract from the input VCF
--chrs [Optional] Chromosomes to use. If not given, then only use chr1~22
--dipcall [Optional] Flag for VCF generated by Dipcall (due to the format difference) 
--remove_small_sv [Optional] Flag for removing variants below 50bp in the output VCFs
--passonly []
```
