# SV Evaluators
**Special Note:** The benchmark VCFs and high confidence BED file are provided [here](https://github.com/maiziezhoulab/LRSV_combo/tree/main/benchmarks) 
# Table of Contents
- [Truvari](#Truvari)
- [hap-eval](#hap-eval)
- [SURVIVOR](#SURVIVOR)


# Truvari
## Project Links
### Github Repo:
https://github.com/ACEnglish/truvari
### Publication:
Truvari: refined structural variant comparison preserves allelic diversity

https://doi.org/10.1186/s13059-022-02840-6
#### BibTeX
```
@article{english2022truvari,
  title={Truvari: refined structural variant comparison preserves allelic diversity},
  author={English, Adam C and Menon, Vipin K and Gibbs, Richard A and Metcalf, Ginger A and Sedlazeck, Fritz J},
  journal={Genome Biology},
  volume={23},
  number={1},
  pages={1--20},
  year={2022},
  publisher={Springer}
}
```
## Installation & Dependencies
### Installation Methods
See [here](https://github.com/acenglish/truvari/wiki/Installation)
### Dependencies
```
Not explicilty mentioned. Handled by Conda/Mamba/pip
```
## Commands used
```

```
## Other notes
1.For more detailed usage, please refer to their [wiki page](https://github.com/ACEnglish/truvari/wiki)

# hap-eval
## Project Links
### Github Repo:

### Publication:

#### BibTeX
```

```
## Installation & Dependencies
### Installation Methods

### Dependencies
```

```
## Inputs & Outputs
### Inputs

### Outputs

## Commands used
```

```
## Other notes
### Pipeline and parameter definition

# SURVIVOR
## Project Links
### Github Repo:

### Publication:

#### BibTeX
```

```
## Installation & Dependencies
### Installation Methods

### Dependencies
```

```
## Inputs & Outputs
### Inputs

### Outputs

## Commands used
```

```
## Other notes
